src/wp-includes/html-api/class-wp-html-processor.php

```diff
@@ -1752,7 +1752,19 @@ private function step_in_table(): bool {
 			 * > A character token, if the current node is table, tbody, template, tfoot, thead, or tr element
 			 */
 			case '#text':
-				return $this->step_in_table_text();
+				$current_token = $this->bookmarks[ $this->state->current_token->bookmark_name ];
+				if (
+					strspn( $this->html, "\u{0000}\u{0009}\u{000A}\u{000C}\u{000D}\u{0020}", $current_token->start, $current_token->length ) === $current_token->length
+				) {
+					if ( strspn( $this->html, "\u{0000}", $current_token->start, $current_token->length ) === $current_token->length ) {
+						// @todo Indicate a parse error once it's possible.
+						return $this->step();
+					}
+					$this->insert_html_element( $this->state->current_token );
+					return true;
+				}
+				$this->bail( 'Foster parenting is not supported.' );
```

Comment:
can we just say that this parser doesn't support foster parenting and be good with it? I don't have as good of a bearing on this as I do with SCRIPT support, where it's clear that the server-side parser can't support it.

foster parenting impacts structure and CSS selectors. what would happen if implemented the rules but foster parenting were always disabled? 🤔

Reply:
[To be fully compliant, I don't think that is allowed](https://html.spec.whatwg.org/#creating-and-inserting-nodes):

> While the parser is processing a token, it can enable or disable foster parenting. This affects the following algorithm.
> *Note* Foster parenting happens when content is misnested in tables.

Are you concerned about proceeding with this change while foster parenting is not supported, or is the question about whether we need to support it in the long term?

Reply:
> foster parenting impacts structure and CSS selectors. what would happen if implemented the rules but foster parenting were always disabled?

If it continued to bail, then we'd refuse to process HTML with misnested content in tables. This would be a limitation and a deviation from the specification, where HTML with a valid representation would be rejected by the HTML Processor.

In order to support foster parenting, I think we'd need a design shift that would likely be harmful to general performance and memory overhead in certain circumstances. When reaching a table, I believe we'd need to _process the entire table_ before pausing on the table node because of cases like this:

```html
<table>
  <!-- A gigantic valid table -->
  <span>Oops! In HTML this appears at the very end is actually before the table!</span>
</table>
```

If we expect to reach those nodes in order, the `SPAN` is the sibling of the `TABLE` immediately preceding it, so `SPAN` should be reached first.

If we ignore ordering, then matching CSS selectors (depending on the supported selectors) we could recognize that the `SPAN` is not a descendant of the TABLE, but of the TABLE's parent.

Reply:
> Are you concerned about proceeding with this change while foster parenting is not supported, or is the question about whether we need to support it in the long term?

Long term - not about this PR

Reply:
> When reaching a table, I believe we'd need to process the entire table before pausing on the table node because of cases like this:

I agree, which is why I've always wanted it to be opt-in.

But on the other hand, what if we just say _this is not a foster-parenting-enabled parser_ and note that as a limitation? That may not be a divergence from the spec because we have chosen to build a parser without foster parenting enabled.

Reply:
> what if we just say this is not a foster-parenting-enabled parser and note that as a limitation

That's certainly viable in the short term, the HTML Processor does not support everything now. We don't need to finalize the decision of whether to support fostering right now, it can be deferred until later.

There may also be a good middle ground where the processor could try to parse some of the table to look for fostering, then bail after a threshold is reached.

**For now** I think we should certainly note the fostering is not enabled (and explain what that means). If and when it makes sense, we can add support later.

Reply:
well I think you are answering a different question than I asked, but I believe that I found my answer in the spec and now I remember getting to this point before.

the question is whether enabling foster parenting is a choice of the parser or not. apparently it's not, and the toggling of foster parenting is triggered in the [anything else](https://html.spec.whatwg.org/#parsing-main-intable:foster-parent) rule of `IN TABLE` insertion mode.

I keep hoping this is more like the scripting tag, or the speculative parser, or whether we report errors and the decision to allow foster parenting is a choice of the parser. alas. we'll just carry on as usual.